### PR TITLE
Re-architecture of UpholdJob, BitflyerJob, and GeminiJob to use shared

### PR DESCRIPTION
**Pull request by yachtcaptain23**
Re-architecture of UpholdJob, BitflyerJob, and GeminiJob to use shared logic and to handle restarts in case if enqueueing payouts fails.

This implementation is currently meant to be run single threaded, and might need refactoring to use less memory in the :pluck field. However, given that uuid has 32 characters, we likely won't hit on-device memory constraints anytime soon.

Closes https://github.com/brave-intl/publishers/issues/3286

### CODE REVIEW

File: app/jobs/enqueue_publishers_for_payout_job.rb [hunk 1]
```diff
@@ -38,29 +38,27 @@ def perform(should_send_notifications: false, final: true, manual: false, payout
...
-    )
-    Payout::GeminiJob.perform_later(
-      payout_report_id: payout_report.id,
-      publisher_ids: publisher_ids
-    )
-    Payout::BitflyerJob.perform_later(
-      payout_report_id: payout_report.id,
-      publisher_ids: publisher_ids
-    )
+    }.to_json
```
Comment (tsmartt) [inline app/jobs/enqueue_publishers_for_payout_job.rb:58]: Nice.

File: app/jobs/payout/concerns/retries_from_last_enqueued_publisher.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
+module RetriesFromLastEnqueuedPublisher
+  extend ActiveSupport::Concern
+
+  def enqueue_retryable_potential_payout!
```
Comment (tsmartt) [inline app/jobs/payout/concerns/retries_from_last_enqueued_publisher.rb:4]: 👏  for making the 3 classes 1. Nice reduction in code. Going to a concern here is a problem as it spreads variables across files and makes it harder to reason about. Please extract this to a service class to handle this logic.

File: test/jobs/payout/uphold_job_test.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  test "for order and key presence" do
+    Sidekiq::Worker.clear_all
+    uphold_job = Payout::UpholdJob.new
+    uphold_job.perform(json_args: { payout_report_id: payout_reports(:one).id }.to_json)
+    visited_publisher_id = nil
+    assert IncludePublisherInPayoutReportJob.jobs.count > 1
+    IncludePublisherInPayoutReportJob.jobs.each do |job|
+      job_publisher_id = job["args"].first["publisher_id"]
+      assert visited_publisher_id <= job_publisher_id if visited_publisher_id.present?
```
Comment (tsmartt) [inline test/jobs/payout/uphold_job_test.rb:29]: What if the `visited_publisher_id` is never present?

File: app/jobs/payout/concerns/retries_from_last_enqueued_publisher.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    # self.class.name is the name of the caller function
+    last_enqueued_publisher_key = "#{self.class.name}-#{@payout_report_id || ""}"
```
Comment (tsmartt) [inline app/jobs/payout/concerns/retries_from_last_enqueued_publisher.rb:6]: Can you write up a section about what the technical underlying issue is and why this method fixes it?

**Review by tsmartt — CHANGES_REQUESTED** [review]

File: test/jobs/payout/uphold_job_test.rb (hunk 3 above)
Comment (yachtcaptain23) [reply test/jobs/payout/uphold_job_test.rb:29]: Test checks it at the end

File: app/jobs/payout/concerns/retries_from_last_enqueued_publisher.rb (hunk 2 above)
Comment (yachtcaptain23) [reply app/jobs/payout/concerns/retries_from_last_enqueued_publisher.rb:4]: Sure, I'll move it to a Service.

File: test/jobs/payout/uphold_job_test.rb (hunk 3 above)
Comment (tsmartt) [reply test/jobs/payout/uphold_job_test.rb:29]: It is compared to something which might be nil.

File: app/jobs/payout/service/retries_from_last_enqueued_publisher.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        number_of_payments = PayoutReport.expected_num_payments(publishers)
+        payout_report.with_lock do
+          payout_report.reload
+          payout_report.expected_num_payments = number_of_payments + payout_report.expected_num_payments
+          payout_report.save!
+        end
+      end
+    end
+
+    publishers.pluck(:id).each do |publisher_id|
```
Comment (tsmartt) [inline app/jobs/payout/service/retries_from_last_enqueued_publisher.rb:37]: maybe use [in_batches ](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-in_batches)here?

File: test/jobs/payout/uphold_job_test.rb (hunk 3 above)
Comment (yachtcaptain23) [reply test/jobs/payout/uphold_job_test.rb:29]: It's checked at the end of the line with `if visited_publisher_id.present?`

File: app/jobs/payout/service/retries_from_last_enqueued_publisher.rb (hunk 5 above)
Comment (yachtcaptain23) [reply app/jobs/payout/service/retries_from_last_enqueued_publisher.rb]: `in_batches` returns a ActiveRecord::Relation, whereas `find_each` handles batching

File: app/jobs/payout/concerns/retries_from_last_enqueued_publisher.rb (hunk 4 above)
Comment (yachtcaptain23) [reply app/jobs/payout/concerns/retries_from_last_enqueued_publisher.rb]: Discussed offline. Resolving

**Review by tsmartt — APPROVED** [review]

**Pull request merged**